File: 문제풀이/2024/052024 백준 Magic Cards 30858 G2.py

from collections import defaultdict
from sys import stdin
input = stdin.readline


def solution(N, K, M, F, cards, queries):
    cards_str = [['N']*K for _ in range(N+1)]

    for card_no, card in enumerate(cards):
        for num in card:
            cards_str[num][card_no] = 'Y'

    memo = defaultdict(int)
    for num in range(1, N+1):
        s = ''.join([cards_str[num][i] for i in range(K)])

        if s in memo:
            memo[s] = 0
        else:
            memo[s] = num

    ret = [memo[query] for query in queries]
    return ret


def main():
    N, K, M, F = map(int, input().split())
    cards = [list(map(int, input().split())) for _ in range(K)]
    queries = [input().rstrip() for _ in range(F)]

    print(*solution(N, K, M, F, cards, queries), sep="\n")

    return 0


if __name__ == '__main__':
    main()

File: 문제풀이/2024/052024 백준 Magic Cards 30858 G2.rs
use std::collections::{HashMap, VecDeque};
use std::io::{self};

fn input() -> VecDeque<Vec<String>> {
    let input = io::read_to_string(io::stdin()).unwrap();
    input
        .trim()
        .split('\n')
        .map(|str| {
            str.split_ascii_whitespace()
                .map(|st| st.to_string())
                .collect::<Vec<String>>()
        })
        .collect()
}

fn main() {
    let mut input = input();

    let line = input
        .pop_front()
        .unwrap()
        .into_iter()
        .map(|st| st.parse::<usize>().unwrap())
        .collect::<Vec<usize>>();
    let (n, k, m, f) = (line[0], line[1], line[2], line[3]);

    let mut cards_str = vec![vec!['N'; k]; n + 1];

    for (card_no, card) in input.drain(..k).enumerate() {
        for num in card {
            cards_str[num.parse::<usize>().unwrap()][card_no] = 'Y';
        }
    }

    let mut memo: HashMap<String, usize> = HashMap::new();

    for num in 1..(n + 1) {
        let s = (0..k)
            .map(|card_no| cards_str[num][card_no])
            .collect::<String>();

        if memo.contains_key(&s) {
            memo.insert(s, 0);
        } else {
            memo.insert(s, num);
        }
    }

    let ret = input
        .drain(..f)
        .map(|query| memo.get(&query.join("")).unwrap())
        .map(|&num| num.to_string())
        .collect::<Vec<String>>();

    println!("{}", ret.join("\n"));
}
